core/arch/x86/decode.c

```diff
@@ -1446,20 +1456,24 @@ decode_reg(decode_reg_t which_reg, decode_info_t *di, byte optype, opnd_size_t o
     case TYPE_VSIB: {
         reg_id_t extend_reg = extend ? reg + 8 : reg;
         extend_reg = avx512_extend ? extend_reg + 16 : extend_reg;
-        return (TEST(PREFIX_EVEX_LL, di->prefixes)
-                    ? (DR_REG_START_ZMM + extend_reg)
-                    : ((TEST(PREFIX_VEX_L, di->prefixes) &&
-                        /* Not only do we use this for VEX .LIG (where raw reg is either
-                         * OPSZ_32 or OPSZ_16_vex32) but also for VSIB which currently
-                         * does not get up to OPSZ_16 so we can use this negative
-                         * check.
-                         * XXX i#1312: vgather/vscatter VSIB addressing may be OPSZ_16?
-                         * For EVEX .LIG, raw reg will be able to be OPSZ_64 or
-                         * OPSZ_16_vex32_evex64.
-                         */
-                        expand_subreg_size(opsize) != OPSZ_16)
-                           ? (REG_START_YMM + extend_reg)
-                           : (REG_START_XMM + extend_reg)));
+        bool operand_is_zmm = TEST(PREFIX_EVEX_LL, di->prefixes) &&
+            expand_subreg_size(opsize) != OPSZ_16 &&
+            expand_subreg_size(opsize) != OPSZ_32;
```

Comment:
Not sure I follow this negative test vs just 2 sizes: what if it returns OPSZ_8?  What if it returns OPSZ_16_vex32?  I don't see how this indicates zmm?

Reply:
This is indeed a little bit fragile and was already before this patch. This relies on the fact that the cases (right now) where EVEX.LL' == 1 and register is _not_ zmm, the expand_subreg_size is OPSZ_16, OPSZ_32 or OPSZ_vex32_evex64 (expand_subreg_size is OPSZ_32). I don't like this code myself. In order to make old and new code really robust, maybe we should add a new type like TYPE_W_128 or TYPE_W_256? Leave it for now as it is well tested, for the sake of forward progress? I have added a XXX comment.

Reply:
Hmm, I would say let's have a comment in expand_subreg_size() too so we don't break this?  If there's an easy way to assert or sthg we could add that but I'm not sure it's easy.

Reply:
Done, added comments. An easy assertion would mean there was an easy better way to detect. I think we should re-factor and possibly even need new types above? We're covered pretty well by testing though I think (disassembler, instruction testing). I am confident it works, though very easy to break.